# 4.7 Deriving New Variables

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\ophab\Desktop\JESSICA\Careerfoundry JEJE\Data Immersion\09-2025 Instacart Basket Analysis'

In [3]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [4]:
df_ords_prods_merge.shape

(32404859, 18)

#### If-statement with user-defined functions

In [5]:
# Define a function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return np.nan

In [6]:
df_ords_prods_merge['price_range'] = df_ords_prods_merge.apply(price_label, axis=1)

In [8]:
df_ords_prods_merge['price_range'].value_counts(dropna = False)

price_range
Mid-range product    21860860
Low-range product    10126321
High range             417678
Name: count, dtype: int64

In [9]:
df_ords_prods_merge['prices'].max()

99999.0

In [11]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 100]

Unnamed: 0_x  order_id  user_id eval_set  order_number  \
1576               171    912404       17    prior            12   
1638               181    603376       17    prior            22   
16522             1944   3264360      135    prior             2   
16528             1945    892534      135    prior             3   
53672             5708    229704      342    prior             8   
...                ...       ...      ...      ...           ...   
32320511       3411856   3172853   205650    prior            18   
32347717       3414691   2504315   205818    prior             3   
32347727       3414693   1108388   205818    prior             5   
32380527       3418470   1916142   206049    prior             1   
32380551       3418473    379732   206049    prior             4   

          orders_day_of_week  order_hour_of_day  days_since_prior_order  \
1576                       2                 14                     5.0   
1638                       6                 16                     4.0   
16522                      2                 21                    13.0   
16528                      0                  8                    12.0   
53672                      1                 19                    30.0   
...                      ...                ...                     ...   
32320511                   1                  9                     7.0   
32347717                   5                 15                     3.0   
32347727                   4                  5                     1.0   
32380527                   2                 17                     NaN   
32380551                   1                 14                     5.0   

          product_id  add_to_cart_order  reordered _merge  Unnamed: 0_y  \
1576           21553                  5          0   both         21554   
1638           21553                  3          1   both         21554   
16522          21553                  6          0   both         21554   
16528          21553                  3          1   both         21554   
53672          21553                  9          0   both         21554   
...              ...                ...        ...    ...           ...   
32320511       21553                 17          1   both         21554   
32347717       21553                 13          0   both         21554   
32347727       21553                  5          1   both         21554   
32380527       21553                  2          0   both         21554   
32380551       21553                  4          1   both         21554   

                              product_name  aisle_id  department_id   prices  \
1576      Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
1638      Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
16522     Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
16528     Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
53672     Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
...                                    ...       ...            ...      ...   
32320511  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
32347717  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
32347727  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
32380527  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
32380551  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   

         merge_flag price_range  
1576           both  High range  
1638           both  High range  
16522          both  High range  
16528          both  High range  
53672          both  High range  
...             ...         ...  
32320511       both  High range  
32347717       both  High range  
32347727       both  High range  
32380527       both  High range  
32380551       both  High range  

[5127 rows x 19 colum

In [13]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [15]:
df_ords_prods_merge['prices'].max()

25.0

In [17]:
df_ords_prods_merge['price_range'].value_counts(dropna = False)

price_range
Mid-range product    21860860
Low-range product    10126321
High range             417678
Name: count, dtype: int64

In [19]:
df_ords_prods_merge['prices'].describe()

count    3.239973e+07
mean     7.790994e+00
std      4.241809e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

#### For Loop

In [20]:
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [21]:
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [22]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy',
 'Least busy

In [23]:
df_ords_prods_merge['busiest_day'] = result

In [24]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# TASK 

#### Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [25]:
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result.append("Busiest days")
  elif value == 3 or value == 4:
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

In [26]:
df_ords_prods_merge["day_category"] = result

In [27]:
df_ords_prods_merge['day_category'].value_counts(dropna = False)

day_category
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [28]:
busiest_sum = 6204182 + 5660230
slowest_sum = 3840534 + 3783802
Regularly_sum = 4496490 + 4213830 + 4205791

In [29]:
print("Busiest days sum:", busiest_sum)
print("Slowest days sum:", slowest_sum)
print("Regularly busy sum:", Regularly_sum)

Busiest days sum: 11864412
Slowest days sum: 7624336
Regularly busy sum: 12916111


I created a new column day_category to label days of the week based on order volume. The two days with the highest orders were labeled "Busiest days", the two days with the lowest orders were labeled "Slowest days", and the remaining days were labeled "Regularly busy".
To double-check the results, I calculated the sum of orders for each category:
Busiest days sum: 11,864,412
Slowest days sum: 7,624,336
Regularly busy sum: 12,944,211
-- The sums match the expected totals, confirming that the labels were applied correctly.
-- This column now provides a clear summary of order activity by day of the week.

#### When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”¶

In [30]:
df_ords_prods_merge.head()

Unnamed: 0_x  order_id  user_id eval_set  order_number  orders_day_of_week  \
0             0   2539329        1    prior             1                   2   
1             0   2539329        1    prior             1                   2   
2             0   2539329        1    prior             1                   2   
3             0   2539329        1    prior             1                   2   
4             0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   ...  _merge Unnamed: 0_y                             product_name aisle_id  \
0  ...    both          195                                     Soda       77   
1  ...    both        14084  Organic Unsweetened Vanilla Almond Milk       91   
2  ...    both        12427                      Original Beef Jerky       23   
3  ...    both        26089               Aged White Cheddar Popcorn       23   
4  ...    both        26406         XL Pick-A-Size Paper Towel Rolls       54   

   department_id  prices  merge_flag        price_range     busiest_day  \
0              7     9.0        both  Mid-range product  Regularly busy   
1             16    12.5        both  Mid-range product  Regularly busy   
2             19     4.4        both  Low-range product  Regularly busy   
3             19     4.7        both  Low-range product  Regularly busy   
4             17     1.0        both  Low-range product  Regularly busy   

     day_category  
0  Regularly busy  
1  Regularly busy  
2  Regularly busy  
3  Regularly busy  
4  Regularly busy  

[5 rows x 21 columns]

In [31]:
orders_per_hour = df_ords_prods_merge ['order_hour_of_day'].value_counts()

In [32]:
print(orders_per_hour)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64


In [33]:
top_threshold = orders_per_hour.quantile(0.75)
bottom_threshold = orders_per_hour.quantile(0.25)

In [34]:
busiest_period = []

for hour in df_ords_prods_merge["order_hour_of_day"]:
    orders_count = orders_per_hour[hour]  # get total orders for this hour
    if orders_count >= top_threshold:
        busiest_period.append("Most orders")
    elif orders_count <= bottom_threshold:
        busiest_period.append("Fewest orders")
    else:
        busiest_period.append("Average orders")

In [37]:
df_ords_prods_merge["busiest_period_of_day"] = busiest_period

In [38]:
busiest_period

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average or

In [39]:
df_ords_prods_merge["busiest_period_of_day"].value_counts()

busiest_period_of_day
Most orders       16128666
Average orders    15679865
Fewest orders       596328
Name: count, dtype: int64

I created a new column, busiest_period_of_day, to classify hours of the day based on order volume. Initially, I made an indexed frequency table to pick the top 5 and bottom 5 hours, but this method seemed too subjective and potentially biased. After conducting research, I remembered the quartile concept from Excel and looked for a similar solution in Python. Rather than picking the busiest and slowest hours manually, I decided to use percentiles to make the classification more systematic and data-driven. The top 25% of hours by order count were labeled "Most orders", the bottom 25% were labeled "Fewest orders", and the remaining hours were labeled "Average orders". This approach ensures that the labeling adapts to the actual distribution of orders, avoids arbitrary decisions, and provides a clear overview of peak, average, and low-activity periods. I also checked the resulting counts for each category to confirm the labeling was applied correctly.¶

In [41]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_new_variables_REVISED.pkl'))